apps/hash-graph/lib/graph/src/snapshot.rs

```diff
@@ -0,0 +1,156 @@
+    ontology::{
+        CustomOntologyMetadata, OntologyTemporalMetadata, OntologyTypeMetadata, OntologyTypeRecord,
+    },
+};
+pub use crate::snapshot::metadata::SnapshotMetadata;
+use crate::{
+    knowledge::Entity,
+    store::{crud::Read, query::Filter, InsertionError, QueryError},
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum SnapshotEntry {
+    Snapshot(SnapshotMetadata),
+    DataType(OntologyTypeRecord<DataType>),
+    PropertyType(OntologyTypeRecord<PropertyType>),
+    EntityType(OntologyTypeRecord<EntityType>),
+    Entity(EntityRecord),
+}
+
+pub struct SnapshotStore<S>(S);
+
+impl<S> SnapshotStore<S> {
+    pub const fn new(store: S) -> Self {
+        Self(store)
+    }
+}
+
+#[expect(
+    clippy::trait_duplication_in_bounds,
+    reason = "False positive: the generics are different"
+)]
+impl<S> SnapshotStore<S>
+where
+    S: Read<OntologyTypeRecord<DataType>>
+        + Read<OntologyTypeRecord<PropertyType>>
+        + Read<OntologyTypeRecord<EntityType>>
+        + Read<Entity>
+        + Send,
+{
+    /// Reads the snapshot from the store into the given sink.
+    ///
+    /// The sink is expected to be a `futures::Sink` that can be used to write the snapshot entries
+    /// into.
+    ///
+    /// # Errors
+    ///
+    /// - If reading a record from the datastore fails
+    /// - If writing a record into the sink fails
+    pub async fn dump_snapshot(
+        &self,
+        sink: impl Sink<SnapshotEntry, Error = Report<impl Context>> + Send,
+    ) -> Result<(), QueryError> {
+        futures::stream::once(async move {
+            Ok(SnapshotEntry::Snapshot(SnapshotMetadata {
+                block_protocol_module_versions: BlockProtocolModuleVersions {
+                    graph: semver::Version::new(0, 3, 0),
+                },
+                custom: CustomGlobalMetadata,
+            }))
+        })
+        .chain(
+            Read::<OntologyTypeRecord<DataType>>::read(&self.0, &Filter::All(vec![]), None)
+                .await?
+                .map_ok(SnapshotEntry::DataType),
+        )
+        .chain(
+            Read::<OntologyTypeRecord<PropertyType>>::read(&self.0, &Filter::All(vec![]), None)
+                .await?
+                .map_ok(SnapshotEntry::PropertyType),
+        )
+        .chain(
+            Read::<OntologyTypeRecord<EntityType>>::read(&self.0, &Filter::All(vec![]), None)
+                .await?
+                .map_ok(SnapshotEntry::EntityType),
+        )
+        .chain(
+            Read::<Entity>::read(&self.0, &Filter::All(vec![]), None)
+                .await?
+                .map_ok(|entity| SnapshotEntry::Entity(entity.into())),
+        )
+        .forward(sink.sink_map_err(|report| {
+            report
+                .change_context(QueryError)
+                .attach_printable("failed to write record into sink")
```

Comment:
Do we want to make this typed? 

I think we should lower our usage of plain strings, these sorts of messages don't end up crossing API boundaries

Reply:
Yes, I think it's good to have it from the start, thanks for flagging! [Improve error reporting for snapshot store](https://github.com/hashintel/hash/pull/2355/commits/a1b25241a3fc0ace4a0462854c9b28e68cc20d26)